lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -9,445 +9,357 @@
-      if (type_isa<RefType>(argType))
-        continue;
-      if (module.getPortDirection(arg.getArgNumber()) == Direction::In)
-        worklist.push_back(arg);
-      if (!argType.isGround())
-        setValRefsTo(arg, FieldRef(arg, 0));
-    }
-    DenseSet<Value> memPorts;
-
-    for (auto &op : module.getOps()) {
-      TypeSwitch<Operation *>(&op)
-          // Wire is added to the worklist
-          .Case<WireOp>([&](WireOp wire) {
-            worklist.push_back(wire.getResult());
-            auto ty = type_dyn_cast<FIRRTLBaseType>(wire.getResult().getType());
-            if (ty && !ty.isGround())
-              setValRefsTo(wire.getResult(), FieldRef(wire.getResult(), 0));
+  void constructReachingDefGraph(FModuleOp module) {
+    LLVM_DEBUG(llvm::dbgs() << "\n Module :" << module.getName());
+    for (auto &op : module) {
+      llvm::TypeSwitch<Operation *>(&op)
+          .Case<FConnectLike>([&](FConnectLike connect) {
+            auto src = connect.getSrc();
+            auto dst = connect.getDest();
+            recordDataflow(dst, src);
           })
-          // All sub elements are added to the worklist.
-          .Case<SubfieldOp>([&](SubfieldOp sub) {
-            auto res = sub.getResult();
-            bool isValid = false;
-            auto fieldIndex = sub.getAccessedField().getFieldID();
-            if (memPorts.contains(sub.getInput())) {
-              auto memPort = sub.getInput();
-              auto type = memPort.getType();
-              auto enableFieldId =
-                  type.getFieldID((unsigned)ReadPortSubfield::en);
-              auto dataFieldId =
-                  type.getFieldID((unsigned)ReadPortSubfield::data);
-              auto addressFieldId =
-                  type.getFieldID((unsigned)ReadPortSubfield::addr);
-              if (fieldIndex == enableFieldId || fieldIndex == dataFieldId ||
-                  fieldIndex == addressFieldId) {
-                setValRefsTo(memPort, FieldRef(memPort, 0));
-              } else
-                return;
-            }
-            SmallVector<FieldRef, 4> fields;
-            forallRefersTo(
-                sub.getInput(),
-                [&](FieldRef subBase) {
-                  isValid = true;
-                  fields.push_back(subBase.getSubField(fieldIndex));
-                  return success();
-                },
-                false);
-            if (isValid) {
-              for (auto f : fields)
-                setValRefsTo(res, f);
-            }
+          .Case<NodeOp>([&](NodeOp op) {
+            auto src = op.getInput();
+            auto dst = op.getData();
+            recordDataflow(dst, src);
+            if (op.isForceable())
+              recordDataflow(op.getRef(), src);
+          })
+          .Case<MemOp>([&](MemOp mem) { handleMemory(mem); })
+          .Case<WireOp>([&](WireOp wire) {
+            if (wire.isForceable())
+              recordDataflow(wire.getData(), wire.getRef());
```

Comment:
You are right, merging the nodes makes more sense, I was thinking the ref as remote write, but it can also introduce cycles with remote reads. Will update the logic.